airbyte-webapp/src/components/base/Text/Text.tsx

```diff
@@ -0,0 +1,68 @@
+import classNames from "classnames";
+import React from "react";
+
+import headingStyles from "./heading.module.scss";
+import textStyles from "./text.module.scss";
+
+type TextSize = "sm" | "md" | "lg";
+type HeadingSize = TextSize | "xl";
+type TextElementType = "p" | "span" | "div";
+type HeadingElementType = "h1" | "h2" | "h3" | "h4" | "h5" | "h6";
+
+interface BaseProps {
+  className?: string;
+  centered?: boolean;
+}
+
+interface TextProps extends BaseProps {
+  as?: TextElementType;
+  size?: TextSize;
+  bold?: boolean;
+}
+
+interface HeadingProps extends BaseProps {
+  as?: HeadingElementType;
+  size?: HeadingSize;
+}
+
+const getTextClassNames = ({ size, centered, bold }: Required<Pick<TextProps, "size" | "centered" | "bold">>) => {
+  const sizes: Record<TextSize, string> = {
+    sm: textStyles.sm,
+    md: textStyles.md,
+    lg: textStyles.lg,
+  };
+
+  return classNames(textStyles.text, sizes[size], centered && textStyles.centered, bold && textStyles.bold);
+};
+
+const getHeadingClassNames = ({ size, centered }: Required<Pick<HeadingProps, "size" | "centered">>) => {
+  const sizes: Record<HeadingSize, string> = {
+    sm: headingStyles.sm,
+    md: headingStyles.md,
+    lg: headingStyles.lg,
+    xl: headingStyles.xl,
+  };
+
+  return classNames(headingStyles.heading, sizes[size], centered && headingStyles.centered);
+};
+
+const isHeadingType = (props: TextProps | HeadingProps): props is HeadingProps => {
+  return props.as ? /^h[0-6]$/.test(props.as) : false;
+};
+
+export const Text: React.FC<TextProps | HeadingProps> = React.memo((props) => {
+  const isHeading = isHeadingType(props);
+  const { as = isHeading ? "h1" : "p", centered = false, children } = props;
```

Comment:
```
as = isHeading ? "h1" : "p"
```
This was slightly confusing to me, because `isHeading` will only ever be true if `props.as` is set, and its value is like `hX`. So it seems like we don't need to have a condition in the default value of `as` for when `isHeading` is true, because if it is true then we know `props.as` is set, so the default value will not be used.

Therefore, maybe it make sense to just set the default value here to `p`? Lmk if I'm misunderstanding anything

Reply:
This is a good point. Probably Edmundo just coding defensively!

Reply:
Good catch. Typing is a bit tricky here if we don't check for the heading type, that's why it partially destructuring the props. However, I realized that this is not needed and we can indeed assign 'p' to it.